File: backend/src/application/dtos/HouseDTO.ts
import 'reflect-metadata';
import {
  IsNotEmpty,
  IsString,
  IsNumber,
  IsEnum,
  IsBoolean,
  ValidateNested,
  ArrayMinSize,
  IsDefined,
  IsNotEmptyObject,
  IsObject,
  IsBooleanString,
} from "class-validator";


import { HouseCategory } from "domain/types/types";
import { LocationDTO } from "./LocationDTO";
import { Transform, TransformFnParams, Type } from "class-transformer";

export class HouseDTO {
  constructor(props: HouseDTO) {
    Object.assign(this, props, { location: new LocationDTO(props.location)});
  }

  @IsNotEmpty({ message: "Title is required" })
  @IsString({ message: "Title must be a string" })
  title: String;

  @IsNotEmpty({ message: "Min price is required" })
  @IsNumber({}, { message: "Min price must be a Number" })
  minPrice: Number;

  @IsNotEmpty({ message: "Max price is required" })
  @IsNumber({}, { message: "Max price must be a Number" })
  maxPrice: Number;

  @IsNotEmpty({ message: "Category is required" })
  @IsEnum(HouseCategory, { message: "Invalid house category" })
  category: HouseCategory;

  @IsNotEmpty({ message: "Rooms is required" })
  @IsNumber({}, { message: "Rooms must be a Number" })
  rooms: Number;

  @IsNotEmpty({ message: "Beds is required" })
  @IsNumber({}, { message: "Beds must be a Number" })
  beds: Number;

  @IsNotEmpty({ message: "Baths is required" })
  @IsNumber({}, { message: "Baths must be a Number" })
  baths: Number;

  @IsNotEmpty({ message: "Kitchens is required" })
  @IsNumber({}, { message: "Kitchens must be a Number" })
  kitchens: Number;

  @IsNotEmpty({ message: "Size is required" })
  @IsNumber({}, { message: "Size must be a Number" })
  size: Number;

  @IsNotEmpty({ message: "Size unit is required" })
  @IsString({ message: "Size unit must be a string" })
  sizeUnit: String;

  @IsObject()
  @IsNotEmptyObject()
  @ValidateNested()
  @Type(() => LocationDTO)
  location: LocationDTO;

  // @ArrayMinSize(0, { message: "At least zero other feature is required" })
  otherFeatures: String[];

  @IsNotEmpty({ message: "Description is required" })
  @IsString({ message: "Description must be a string" })
  description: String;

  @Transform(({ value }: TransformFnParams) => value?.trim())
  @IsBoolean({ message: "isApproved must be a boolean"})
  isApproved: Boolean;
}

File: mobile/dalali_hub/lib/data/remote/model/house_dto.dart
import 'dart:io';

import 'package:dalali_hub/data/remote/model/location_dto.dart';
import 'package:dalali_hub/domain/entity/house.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'house_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class HouseDto {
  final String? id;
  final String title;
  final double minPrice;
  final double maxPrice;
  final double rooms;
  final double beds;
  final double baths;
  final double kitchens;
  final double size;
  final String sizeUnit;
  final List<String> otherFeatures;
  final String description;
  final bool isApproved;
  final String category;
  final Map location;

  HouseDto({
    required this.title,
    required this.category,
    required this.minPrice,
    required this.maxPrice,
    required this.rooms,
    required this.beds,
    required this.baths,
    required this.kitchens,
    required this.size,
    required this.sizeUnit,
    required this.otherFeatures,
    required this.description,
    required this.isApproved,
    this.id,
    required this.location,
  });
  Map<String, dynamic> toJson() => _$HouseDtoToJson(this);

 static HouseDto toHouseDto(House house) {
    return HouseDto(
      id: house.id,
      title: house.title,
      category: house.category,
      minPrice: house.minPrice,
      maxPrice: house.maxPrice,
      rooms: house.rooms,
      beds: house.beds,
      baths: house.baths,
      kitchens: house.kitchens,
      size: house.size,
      sizeUnit: house.sizeUnit,
      otherFeatures: house.otherFeatures,
      description: house.description,
      isApproved: house.isApproved,
      location: {
        "region": house.location.region,
        "district": house.location.district,
        "ward": house.location.ward
      },
    );
  }
}
